lightly/loss/regularizer/co2.py

```diff
@@ -0,0 +1,131 @@
+
+        Args:
+            out0:
+                Output projections of the first set of transformed images.
+            out1:
+                Output projections of the second set of transformed images.
+            negatives:
+                Negative samples to compare against. If this is None, the second
+                batch of images will be used as negative samples.
+
+        Returns:
+            Log probability that a positive samples will classify each negative 
+            sample as the positive sample.
+
+        """
+        batch_size, _ = out0.shape
+        if negatives is None:
+            # use second batch as negative samples
+            l_pos = torch.einsum('nc,nc->n', [out0, out1]).unsqueeze(-1)
+            l_neg = torch.einsum('nc,ck->nk', [out0, out1.t()])
+            # remove elements on the diagonal
+            l_neg = l_neg.masked_select(
+                ~torch.eye(batch_size, dtype=bool, device=l_neg.device)
+            ).view(batch_size, batch_size - 1)
+        else:
+            # use memory bank as negative samples
+            negatives = negatives.to(out0.device)
+            l_pos = torch.einsum('nc,nc->n', [out0, out1]).unsqueeze(-1)
+            l_neg = torch.einsum('nc,ck->nk', [out0, negatives.clone().detach()])
+            
+        # concatenate such that positive samples are at index 0
+        logits = torch.cat([l_pos, l_neg], dim=1)
+        # divide by temperature
+        logits = logits / self.t_consistency
+
+        # the input to kl_div is expected to be log(p) and we set the
+        # flag log_target to True, so both probabilities should be passed as log
+        log_probs = torch.nn.functional.log_softmax(logits, dim=-1)
+        return log_probs
+
+
+    def forward(self,
+                out0: torch.Tensor,
+                out1: torch.Tensor):
+        """Computes the CO2 regularization term for two model outputs.
+
+        Args:
+            out0:
+                Output projections of the first set of transformed images.
+            out1:
+                Output projections of the second set of transformed images.
+
+        Returns:
+            The regularization term multiplied by the weight factor alpha.
+
+        """
+
+        # normalize the output to length 1
+        out0 = torch.nn.functional.normalize(out0, dim=1)
+        out1 = torch.nn.functional.normalize(out1, dim=1)
+
+        # ask memory bank for negative samples and extend it with out1 if 
+        # out1 requires a gradient, otherwise keep the same vectors in the 
+        # memory bank (this allows for keeping the memory bank constant e.g.
+        # for evaluating the loss on the test set)
+        out1, negatives = \
+            super(CO2Regularizer, self).forward(out1, update=True)
+        
+        # get log probabilities
+        p = self._get_pseudo_labels(out0, out1, negatives)
+        q = self._get_pseudo_labels(out1, out0, negatives)
```

Comment:
`self._get_pseudo_labels()` is only used with negatives, thus we did you implement it also without negatives?

Reply:
The negatives can be `None` if `memory_bank_size` is set to 0. I will extend the comment.